Speech2TextTransformer

patil-suraj opened the pull request:
# What does this PR do?

This PR adds the S2T model from [fairseq](https://github.com/pytorch/fairseq/tree/master/examples/speech_to_text) for end-to-end ASR and Speech-Translation (ST).

The model architecture is somewhat similar to the mBART model, except
- the encoder contains the convolutional subsampling module to downsample the speech features.
- no token embeddings in encoder.

This PR also adds the `Speech2TextFeatureExtractor`, and `Speech2TextProcessor` classes analogous to `Wav2Vec2` extractor and processor.

The `Speech2TextFeatureExtractor` here has an extra dependency on `torchaudio` which is required for extracting fbank features

The `generate` method works out-of-the-box for S2T!  Usage example 

```python
 import torch
 from transformers import Speech2TextForConditionalGeneration, Speech2TextProcessor
 from datasets import load_dataset
 import soundfile as sf

 def map_to_array(batch):
     speech, _ = sf.read(batch["file"])
     batch["speech"] = speech
     return batch

ds = load_dataset("patrickvonplaten/librispeech_asr_dummy", "clean", split="validation")
samples = ds.map(map_to_array)[5:8]

model = Speech2TextForConditionalGeneration.from_pretrained("valhalla/s2t_librispeech_small")
processor = Speech2TextProcessor.from_pretrained("valhalla/s2t_librispeech_small")

features = processor(samples["speech"], sampling_rate=16_000, padding=True, return_tensors="pt")
gen_tokens = model.generate(
    input_ids=features["input_features"],
    attention_mask=features["attention_mask"],
)
generated = processor.batch_decode(gen_tokens, skip_special_tokens=True)
```

TODOs:
- [x] add tests
- [x] implement `Speech2TextProcessor` after #10324 is merged
- [x] finish docs
- [ ] port and eval the CoVoST2 and MuSTc checkpoints
- [ ] add training/fine-tuning script in a follow-up PR

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -386,8 +386,7 @@ def _prepare_decoder_input_ids_for_generation(
         decoder_input_ids = (
-            torch.ones((input_ids.shape[0], 1), dtype=input_ids.dtype, device=input_ids.device)
-            * decoder_start_token_id
+            torch.ones((input_ids.shape[0], 1), dtype=torch.long, device=input_ids.device) * decoder_start_token_id
```
Great change! There is actually no use-case that the decoder_input_ids is not of type torch.long . I like it :-)

patrickvonplaten commented on src/transformers/models/auto/modeling_auto.py:
```diff
@@ -195,6 +191,13 @@
 )
+
+# Add modeling imports here
+# Add modeling imports here
```
I think there should be just one `# Add modeling imports here`

patrickvonplaten commented on src/transformers/models/speech_to_text_transformer/modeling_speech_to_text_transformer.py:
```diff
@@ -0,0 +1,1380 @@
+]
+
+
+def shift_tokens_right(input_ids: torch.Tensor, pad_token_id: int, decoder_start_token_id: int):
```
again we should add all the `# Copied from statements`

patrickvonplaten commented on src/transformers/models/speech_to_text_transformer/modeling_speech_to_text_transformer.py:
```diff
@@ -0,0 +1,1380 @@
+        self.embedding_dim = embedding_dim
+        self.padding_idx = padding_idx
+        self.weights = self.get_embedding(num_positions + self.offset, embedding_dim, padding_idx)
+        self.register_buffer("_float_tensor", torch.FloatTensor(1))
```
do we need to save a tensor for this? this is just a scalar no?

patrickvonplaten commented on src/transformers/models/speech_to_text_transformer/modeling_speech_to_text_transformer.py:
```diff
@@ -0,0 +1,1380 @@
+        self.register_buffer("_float_tensor", torch.FloatTensor(1))
+
+    @staticmethod
+    def get_embedding(num_embeddings: int, embedding_dim: int, padding_idx: Optional[int] = None):
```
let's remove the padding_idx

patrickvonplaten replied:
@patil-suraj, I do think we can and should delete the padding_idx argument here since it's just used to zero out a position which it should not do IMO

patrickvonplaten commented on src/transformers/models/speech_to_text_transformer/modeling_speech_to_text_transformer.py:
```diff
@@ -0,0 +1,1380 @@
+        Computes the output length of the convolutional layers
+        """
+
+        for i in range(self.config.num_conv_layers):
```
Let's replace this: `(input_lengths.float() - 1) / 2 + 1).floor()` by `(input_lengths - 1) // 2 + 1` 

In general `a//b` is preferred over `int(a/b)` -> it's cheaper

patrickvonplaten commented on src/transformers/models/speech_to_text_transformer/modeling_speech_to_text_transformer.py:
```diff
@@ -0,0 +1,1380 @@
+        self.max_source_positions = config.max_position_embeddings
+        self.embed_scale = math.sqrt(embed_dim) if config.scale_embedding else 1.0
+
+        self.subsample = Conv1dSubsampler(config)
```
think `conv` would be a better name, but let's change that at a later point

patrickvonplaten commented on src/transformers/models/speech_to_text_transformer/modeling_speech_to_text_transformer.py:
```diff
@@ -0,0 +1,1380 @@
+
+        # expand encoder attention mask
+        if encoder_hidden_states is not None and encoder_attention_mask is not None:
+            encoder_attention_mask = self._get_subsampled_encoder_attn_mask(encoder_attention_mask)
```
I like it!

patrickvonplaten commented on src/transformers/models/speech_to_text_transformer/modeling_speech_to_text_transformer.py:
```diff
@@ -0,0 +1,1380 @@
+
+        padding_idx, vocab_size = config.pad_token_id, config.vocab_size
+
+        if config.share_embeds:
```
This doesn't look very clean to me .... -> first not a big fan of the name `shared` => can we rename this to `word_embedding`? Second, we don't need word embeddings for the encoder, no? There should actually be no need at all to define the word embeddings at this level, they should always only be defined in the decoder IMO.

patrickvonplaten commented on utils/check_repo.py:
```diff
@@ -30,6 +30,9 @@
     # models to ignore for not tested
+    "Speech2TextTransformerEncoder",  # Building part of bigger (tested) model.
+    "Speech2TextTransformerDecoder",  # Building part of bigger (tested) model.
+    "Speech2TextTransformerDecoderWrapper",  # Building part of bigger (tested) model.
```
Think those classes: `Speech2TextTransformerDecoderWrapper` can be deleted for now

patrickvonplaten commented on utils/check_repo.py:
```diff
@@ -75,6 +78,9 @@
     # models to ignore for model xxx mapping
+    "Speech2TextTransformerEncoder",
+    "Speech2TextTransformerDecoder",
+    "Speech2TextTransformerDecoderWrapper",
```
Think those classes: `Speech2TextTransformerDecoderWrapper` can be deleted for now

patrickvonplaten commented on src/transformers/models/speech_to_text_transformer/modeling_speech_to_text_transformer.py:
```diff
@@ -0,0 +1,1380 @@
+        )
+
+        # expand embeddings if needed
+        max_pos = self.padding_idx + 1 + seq_len
```
This is very hacky IMO, but you're right @patil-suraj we cannot change it anymore now due to backward compatibility

patil-suraj commented on src/transformers/models/speech_to_text_transformer/modeling_speech_to_text_transformer.py:
```diff
@@ -0,0 +1,1380 @@
+            # zero pad
+            emb = torch.cat([emb, torch.zeros(num_embeddings, 1)], dim=1)
+        if padding_idx is not None:
+            emb[padding_idx, :] = 0
```
Yes, you right that "padding index can be encountered at every position", but here `position_ids` are computed such that all padding tokens have position id equal to 1 (which is the padding_idx for S2T) and all non-pad tokens begin at padding_idx + 1 (i.e they start from 2).

patil-suraj commented on tests/test_feature_extraction_common.py:
```diff
@@ -266,7 +266,7 @@ def test_padding_accepts_tensors_tf(self):
         input_tf = feat_extract.pad(processed_features, padding="longest", return_tensors="tf")[input_name]
 
-        self.assertTrue(abs(input_np.sum() - input_tf.numpy().sum()) < 1e-2)
+        self.assertTrue(abs(input_np.astype(np.float32).sum() - input_tf.numpy().sum()) < 1e-2)
```
This test was failing sometimes for S2T because here `input_np` is of type `float64` but `input_tf.numpy()/input_pt.numpy().sum())` is `float32`, so casting `input_np` to `float32`

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -368,7 +368,7 @@ def _prepare_attention_mask_for_generation(
         if is_pad_token_in_inputs_ids and is_pad_token_not_equal_to_eos_token_id:
             return input_ids.ne(pad_token_id).long()
-        return input_ids.new_ones(input_ids.shape)
+        return input_ids.new_ones(input_ids.shape).long()
```
```suggestion
        return input_ids.new_ones(input_ids.shape, dtype=torch.long)
```

patrickvonplaten commented on src/transformers/models/speech_to_text_transformer/feature_extraction_speech_to_text_transformer.py:
```diff
@@ -0,0 +1,232 @@
+logger = logging.get_logger(__name__)
+
+
+class Speech2TextTransformerFeatureExtractor(PreTrainedFeatureExtractor):
```
Actually thinking a bit more about, maybe just having `Speech2Text....` everything so `Speech2TextFeatureExtractor` is nicer

patrickvonplaten commented on src/transformers/models/speech_to_text_transformer/modeling_speech_to_text_transformer.py:
```diff
@@ -0,0 +1,1346 @@
+        return outputs
+
+
+class Speech2TextTransformerPreTrainedModel(PreTrainedModel):
```
Think just `Speech2TextConfig` is actually a bit nicer -> what do you think @patil-suraj?

patrickvonplaten commented on src/transformers/models/speech_to_text_transformer/modeling_speech_to_text_transformer.py:
```diff
@@ -0,0 +1,1346 @@
+                module.weight.data[module.padding_idx].zero_()
+
+    @property
+    def dummy_inputs(self):
```
Actually, we should remove this completely => this is not needed for PyTorch

patrickvonplaten commented on .circleci/config.yml:
```diff
@@ -77,8 +77,9 @@ jobs:
                   keys:
                       - v0.4-torch_and_tf-{{ checksum "setup.py" }}
                       - v0.4-{{ checksum "setup.py" }}
+            - run: sudo apt-get -y update && sudo apt-get install -y libsndfile1-dev
```
what is this needed for?

patil-suraj replied:
`soundfile` needs `libsndfile1`, and `torchaudio` needs `soundfile`

patil-suraj replied:
`libsndfile1` is not a python package, we need to install it using `apt`. I will document this.

patil-suraj commented on src/transformers/models/speech_to_text_transformer/tokenization_speech_to_text_transformer.py:
```diff
@@ -0,0 +1,258 @@
+        self.spm_file = spm_file
+        self.sp_model = load_spm(spm_file)
+
+        if lang_codes is not None:
+            self.lang_codes = lang_codes
+            self.langs = LANGUAGES[lang_codes]
+            self.lang_tokens = [f"<lang:{lang}>" for lang in self.langs]
+            self.lang_code_to_id = {lang: self.sp_model.PieceToId(f"<lang:{lang}>") for lang in self.langs}
```
here `LANGUAGES` is a `dict` of dataset to languages mapping. and `lang_codes` specifies which dataset's language codes to use. We need to handle it this way because different datasets (MuST-C, CoVoST2) cover different languages,

LysandreJik commented on setup.py:
```diff
@@ -227,7 +228,7 @@ def run(self):
 
 extras["serving"] = deps_list("pydantic", "uvicorn", "fastapi", "starlette")
-extras["speech"] = deps_list("soundfile")
+extras["speech"] = deps_list("soundfile", "torchaudio")
```
You're having issues in your tests because the `speech` category in the setup.py is contained in `testing`. The `speech` directive is installed in every test suite, so by adding `torchaudio` to it, you're installing `torch` + `torchaudio` to every suite.

This also explains why you have a timeout exceeded in your `run_tests_tf`: it runs all tests!

patil-suraj commented:
@patrickvonplaten, @sgugger , @LysandreJik  The PR is now finalized and ready for your review :)

sgugger commented on docs/source/model_doc/speech_to_text.rst:
```diff
@@ -0,0 +1,135 @@
+Inference
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+`Speech2Text` is a speech model that accepts a float tensor of log-mel filter-bank features extracted from the speech
```
```suggestion
Speech2Text is a speech model that accepts a float tensor of log-mel filter-bank features extracted from the speech
```

sgugger commented on docs/source/model_doc/speech_to_text.rst:
```diff
@@ -0,0 +1,135 @@
+
+`Speech2Text` is a speech model that accepts a float tensor of log-mel filter-bank features extracted from the speech
+signal. It's a transformer based seq2seq model, so the transcripts/translations are generated autoregressively. The
+`generate` method can be used for inference.
```
```suggestion
:obj:`generate()` method can be used for inference.
```

sgugger commented on docs/source/model_doc/speech_to_text.rst:
```diff
@@ -0,0 +1,135 @@
+
+The :class:`transformers.Speech2TextFeatureExtractor` class is responsible for extracting the log-mel filter-bank
+features. The :class:`transformers.Speech2Textprocessor` wraps the processor and tokenizer objects and is used to
+extract the features and decode the text. The feature extractor depends on `torchaudio` so make sure to install it
```
```suggestion
extract the features and decode the text. The feature extractor depends on :obj:`torchaudio` so make sure to install it
```

patrickvonplaten commented on docs/source/model_doc/speech_to_text.rst:
```diff
@@ -0,0 +1,135 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+`Speech2Text` is a speech model that accepts a float tensor of log-mel filter-bank features extracted from the speech
+signal. It's a transformer based seq2seq model, so the transcripts/translations are generated autoregressively. The
```
```suggestion
signal. It's a transformer-based seq2seq model, so the transcripts/translations are generated autoregressively. The
```

sgugger commented on docs/source/model_doc/speech_to_text.rst:
```diff
@@ -0,0 +1,135 @@
+The :class:`transformers.Speech2TextFeatureExtractor` class is responsible for extracting the log-mel filter-bank
+features. The :class:`transformers.Speech2Textprocessor` wraps the processor and tokenizer objects and is used to
+extract the features and decode the text. The feature extractor depends on `torchaudio` so make sure to install it
+before importing the extractor or processor.
```
Add how to install it here: `pip install transformers[speech]` and any potential other instructions.

sgugger commented on docs/source/model_doc/speech_to_text.rst:
```diff
@@ -0,0 +1,135 @@
+
+    For multilingual speech translation models, :obj:`eos_token_id` is used as the :obj:`decoder_start_token_id` and
+    the target language id is forced as the first generated token. To force the target language id as the first
+    generated token, pass the `forced_bos_token_id` parameter to the `generate` method. The following example shows how
```
```suggestion
    generated token, pass the :obj:`forced_bos_token_id` parameter to the :obj:`generate()` method. The following example shows how
```

sgugger commented on docs/source/model_doc/speech_to_text.rst:
```diff
@@ -0,0 +1,135 @@
+        >>> translation = processor.batch_decode(generated_ids)
+
+
+See the [model hub](https://huggingface.co/models?filter=speech_to_text) to look for `Speech2Text` checkpoints.
```
```suggestion
See the `model hub <https://huggingface.co/models?filter=speech_to_text>`__ to look for Speech2Text checkpoints.
```

patrickvonplaten commented on docs/source/model_doc/speech_to_text.rst:
```diff
@@ -0,0 +1,135 @@
+signal. It's a transformer based seq2seq model, so the transcripts/translations are generated autoregressively. The
+`generate` method can be used for inference.
+
+The :class:`transformers.Speech2TextFeatureExtractor` class is responsible for extracting the log-mel filter-bank
```
```suggestion
The :class:`~transformers.Speech2TextFeatureExtractor` class is responsible for extracting the log-mel filter-bank
```

no?

sgugger commented on setup.py:
```diff
@@ -227,19 +228,18 @@ def run(self):
 extras["quality"] = deps_list("black", "isort", "flake8")
 
-extras["all"] = extras["tf"] + extras["torch"] + extras["flax"] + extras["sentencepiece"] + extras["tokenizers"]
+extras["all"] = extras["tf"] + extras["torch"] + extras["flax"] + extras["sentencepiece"] + extras["tokenizers"] + extras["speech"]
```
Not sure if it should be in all -> retrieval is not for instance.

sgugger commented on src/transformers/models/speech_to_text/__init__.py:
```diff
@@ -0,0 +1,88 @@
+        "Speech2TextPreTrainedModel",
+    ]
+
+if is_torch_available() and is_torchaudio_available():
+    _import_structure["feature_extraction_speech_to_text"] = ["Speech2TextFeatureExtractor"]
```
This is going to make the object not being available when torch is installed but not torchaudio. The `is_torchaudio_available` should be removed here, in my opinion, and the `Speech2TextFeatureExtractor` should raise an error at init if it's not available.

Also that object is always imported in the main init, so it actually should always be available, or the main init amended.

patrickvonplaten commented on src/transformers/models/auto/modeling_auto.py:
```diff
@@ -500,6 +502,7 @@
     [
         # Model for Seq2Seq Causal LM mapping
         (M2M100Config, M2M100ForConditionalGeneration),
+        (Speech2TextConfig, Speech2TextForConditionalGeneration),
```
I think I'd prefer to not add it here since it's not really a Seq2Seq LM but a Seq2Seq ASR model

sgugger commented on src/transformers/models/speech_to_text/__init__.py:
```diff
@@ -0,0 +1,88 @@
+if is_torch_available() and is_torchaudio_available():
+    _import_structure["feature_extraction_speech_to_text"] = ["Speech2TextFeatureExtractor"]
+
+if is_torch_available() and is_torchaudio_available() and is_sentencepiece_available():
+    _import_structure["processing_speech_to_text"] = ["Speech2TextProcessor"]
```
Same comment here

patrickvonplaten commented on src/transformers/models/speech_to_text/configuration_speech_to_text.py:
```diff
@@ -0,0 +1,200 @@
+        input_feat_per_channel (:obj:`int`, `optional`, defaults to 80):
+            An integer specifying the size of feature vector. This is also the dimentions of log-mel filter-bank
+            features.
+        input_channels (:obj:`int`, `optionl`, defaults to 1):
```
```suggestion
        input_channels (:obj:`int`, `optional`, defaults to 1):
```

sgugger commented on src/transformers/models/speech_to_text/__init__.py:
```diff
@@ -0,0 +1,88 @@
+            Speech2TextPreTrainedModel,
+        )
+
+    if is_torch_available() and is_torchaudio_available():
+        from .feature_extraction_speech_to_text import Speech2TextFeatureExtractor
+
+    if is_torch_available() and is_torchaudio_available() and is_sentencepiece_available():
+        from .processing_speech_to_text import Speech2TextProcessor
```
Same comments here.

sgugger commented on src/transformers/models/speech_to_text/feature_extraction_speech_to_text.py:
```diff
@@ -0,0 +1,233 @@
+        **kwargs
+    ):
+        super().__init__(feature_size=feature_size, sampling_rate=sampling_rate, padding_value=padding_value, **kwargs)
+        self.num_mel_bins = num_mel_bins
```
Raise an ImportError here if `is_torch_audio_available` returns False.

patrickvonplaten commented on src/transformers/models/speech_to_text/configuration_speech_to_text.py:
```diff
@@ -0,0 +1,200 @@
+            just in case (e.g., 512 or 1024 or 2048).
+        num_conv_layers (:obj:`int`, `optional`, defaults to 2):
+            Number of 1D convolutional layers in the conv module.
+        conv_kernel_sizes (:obj:`List[int]`, `optional`, defaults to :obj:`[5, 5]`):
```
I'd prefer to not have mutable objects as default values -> can we change that to a tuple input, similar to: hhttps://github.com/huggingface/transformers/blob/c19c811a2d9446f606562a8a133cbea79e5affce/src/transformers/models/wav2vec2/configuration_wav2vec2.py#L154

See: https://florimond.dev/blog/articles/2018/08/python-mutable-defaults-are-the-source-of-all-evil/#:~:text=Do%20not%20use%20mutable%20default,mutable%20value%20inside%20the%20function.

sgugger commented on src/transformers/models/speech_to_text/modeling_speech_to_text.py:
```diff
@@ -0,0 +1,1348 @@
+        Args:
+            hidden_states (:obj:`torch.FloatTensor`): input to the layer of shape `(seq_len, batch, embed_dim)`
+            attention_mask (:obj:`torch.FloatTensor`): attention mask of size
+                `(batch, 1, tgt_len, src_len)` where padding elements are indicated by very large negative values.
```
```suggestion
                :obj:`(batch, 1, tgt_len, src_len)` where padding elements are indicated by very large negative values.
```

sgugger commented on src/transformers/models/speech_to_text/modeling_speech_to_text.py:
```diff
@@ -0,0 +1,1348 @@
+            attention_mask (:obj:`torch.FloatTensor`): attention mask of size
+                `(batch, 1, tgt_len, src_len)` where padding elements are indicated by very large negative values.
+            layer_head_mask (:obj:`torch.FloatTensor`): mask for attention heads in a given layer of size
+                `(config.encoder_attention_heads,)`.
```
```suggestion
                :obj:`(config.encoder_attention_heads,)`.
```

sgugger commented on src/transformers/models/speech_to_text/modeling_speech_to_text.py:
```diff
@@ -0,0 +1,1348 @@
+        Args:
+            hidden_states (:obj:`torch.FloatTensor`): input to the layer of shape `(seq_len, batch, embed_dim)`
+            attention_mask (:obj:`torch.FloatTensor`): attention mask of size
+                `(batch, 1, tgt_len, src_len)` where padding elements are indicated by very large negative values.
```
```suggestion
                :obj:`(batch, 1, tgt_len, src_len)` where padding elements are indicated by very large negative values.
```

sgugger commented on src/transformers/models/speech_to_text/modeling_speech_to_text.py:
```diff
@@ -0,0 +1,1348 @@
+    ):
+        """
+        Args:
+            hidden_states (:obj:`torch.FloatTensor`): input to the layer of shape `(seq_len, batch, embed_dim)`
```
```suggestion
            hidden_states (:obj:`torch.FloatTensor`): input to the layer of shape :obj:`(seq_len, batch, embed_dim)`
```

sgugger commented on src/transformers/models/speech_to_text/modeling_speech_to_text.py:
```diff
@@ -0,0 +1,1348 @@
+            hidden_states (:obj:`torch.FloatTensor`): input to the layer of shape `(seq_len, batch, embed_dim)`
+            attention_mask (:obj:`torch.FloatTensor`): attention mask of size
+                `(batch, 1, tgt_len, src_len)` where padding elements are indicated by very large negative values.
+            encoder_hidden_states (:obj:`torch.FloatTensor`): cross attention input to the layer of shape `(seq_len, batch, embed_dim)`
```
```suggestion
            encoder_hidden_states (:obj:`torch.FloatTensor`): cross attention input to the layer of shape :obj:`(seq_len, batch, embed_dim)`
```

sgugger commented on src/transformers/models/speech_to_text/modeling_speech_to_text.py:
```diff
@@ -0,0 +1,1348 @@
+                `(batch, 1, tgt_len, src_len)` where padding elements are indicated by very large negative values.
+            encoder_hidden_states (:obj:`torch.FloatTensor`): cross attention input to the layer of shape `(seq_len, batch, embed_dim)`
+            encoder_attention_mask (:obj:`torch.FloatTensor`): encoder attention mask of size
+                `(batch, 1, tgt_len, src_len)` where padding elements are indicated by very large negative values.
```
```suggestion
                :obj:`(batch, 1, tgt_len, src_len)` where padding elements are indicated by very large negative values.
```

sgugger commented on src/transformers/models/speech_to_text/modeling_speech_to_text.py:
```diff
@@ -0,0 +1,1348 @@
+            encoder_attention_mask (:obj:`torch.FloatTensor`): encoder attention mask of size
+                `(batch, 1, tgt_len, src_len)` where padding elements are indicated by very large negative values.
+            layer_head_mask (:obj:`torch.FloatTensor`): mask for attention heads in a given layer of size
+                `(config.encoder_attention_heads,)`.
```
```suggestion
                :obj:`(config.encoder_attention_heads,)`.
```

sgugger commented on src/transformers/models/speech_to_text/modeling_speech_to_text.py:
```diff
@@ -0,0 +1,1348 @@
+            layer_head_mask (:obj:`torch.FloatTensor`): mask for attention heads in a given layer of size
+                `(config.encoder_attention_heads,)`.
+            encoder_layer_head_mask (:obj:`torch.FloatTensor`): mask for encoder attention heads in a given layer of
+                size `(config.encoder_attention_heads,)`.
```
```suggestion
                size :obj:`(config.encoder_attention_heads,)`.
```

patrickvonplaten commented on src/transformers/models/speech_to_text/feature_extraction_speech_to_text.py:
```diff
@@ -0,0 +1,233 @@
+            Whether or not to zero-mean normalize the extracted features.
+        norm_vars (:obj:`bool`, `optional`, defaults to :obj:`True`):
+            Whether or not to unit-variance normalize the extracted features.
+        return_attention_mask (:obj:`bool`, `optional`, defaults to :obj:`False`):
```
Do you think it might more sense to not even allow the user to set this variable and just set it to `True`? Think Wav2Vec2 is a huge exception with regard to not passing the `attention_mask`

patrickvonplaten commented on src/transformers/models/speech_to_text/feature_extraction_speech_to_text.py:
```diff
@@ -0,0 +1,233 @@
+    mean and variance normalization to the extracted features.
+
+    Args:
+        feature_size (:obj:`int`, defaults to 1):
```
This should not default to 1 IMO

patrickvonplaten replied:
Ok yeah just a typo I guess

sgugger commented on src/transformers/models/speech_to_text/modeling_speech_to_text.py:
```diff
@@ -0,0 +1,1348 @@
+SPEECH_TO_TEXT_INPUTS_DOCSTRING = r"""
+    Args:
+        input_features (:obj:`torch.LongTensor` of shape :obj:`(batch_size, fbank_feature_length, num_mel_bins)`):
+            Float values of fbank features extracted from the raw speech waveform. Raw speech waveform can be obtained
+            by loading a `.flac` or `.wav` audio file into an array of type `List[float]` or a `numpy.ndarray`, *e.g.*
+            via the soundfile library (`pip install soundfile`). To prepare the array into `input_features`, the
+            :class:`Speech2TextTokenizer` should be used for extracting the fbank features, padding and conversion into
+            a tensor of type `torch.FloatTensor`. See :meth:`transformers.Speech2TextTokenizer.__call__`
```
Please remember that:
- \`xxx\` does not mean code formatting in rst, you need to use \`\`xxx\`\` or :obj:\`xxx\`
- :class:\`Speech2TextTokenizer\` will not work, you have to write :class:\`~transformers.Speech2TextTokenizer\`

sgugger commented on src/transformers/models/speech_to_text/modeling_speech_to_text.py:
```diff
@@ -0,0 +1,1348 @@
+    ):
+        r"""
+        Args:
+            input_features (:obj:`torch.LongTensor` of shape :obj:`(batch_size, fbank_feature_length, num_mel_bins)`):
+                Float values of fbank features extracted from the raw speech waveform. Raw speech waveform can be
+                obtained by loading a `.flac` or `.wav` audio file into an array of type `List[float]` or a
+                `numpy.ndarray`, *e.g.* via the soundfile library (`pip install soundfile`). To prepare the array into
+                `input_features`, the :class:`Speech2TextTokenizer` should be used for extracting the fbank features,
+                padding and conversion into a tensor of type `torch.FloatTensor`. See
+                :meth:`transformers.Speech2TextTokenizer.__call__`
```
Same comment here.

patrickvonplaten commented on src/transformers/models/speech_to_text/feature_extraction_speech_to_text.py:
```diff
@@ -0,0 +1,233 @@
+            Number of triangular Mel-frequency bins
+        padding_value (:obj:`float`, defaults to 0.0):
+            The value that is used to fill the padding values.
+        do_normalize (:obj:`bool`, `optional`, defaults to :obj:`True`):
```
maybe give more explicit names here? `do_ceptral_normalize`

patrickvonplaten commented on src/transformers/models/speech_to_text/feature_extraction_speech_to_text.py:
```diff
@@ -0,0 +1,233 @@
+            The value that is used to fill the padding values.
+        do_normalize (:obj:`bool`, `optional`, defaults to :obj:`True`):
+            Whether or not to apply utterance-level cepstral mean and variance normalization to extracted features.
+        norm_means (:obj:`bool`, `optional`, defaults to :obj:`True`):
```
maybe `do_mean_normalize` here for conistency and same for param below?

sgugger commented on src/transformers/__init__.py:
```diff
@@ -135,6 +135,13 @@
+    "models.speech_to_text": [
+        "SPEECH_TO_TEXT_PRETRAINED_CONFIG_ARCHIVE_MAP",
+        "Speech2TextConfig",
+        "Speech2TextTokenizer",
+        "Speech2TextFeatureExtractor",
+        "Speech2TextProcessor",
```
Those should not be here.

patrickvonplaten commented on src/transformers/models/speech_to_text/feature_extraction_speech_to_text.py:
```diff
@@ -0,0 +1,233 @@
+        sampling_rate (:obj:`int`, defaults to 16000):
+            The sampling rate at which the audio files should be digitalized expressed in Hertz per second (Hz).
+        num_mel_bins (:obj:`int`, defaults to 80):
+            Number of triangular Mel-frequency bins
```
I'd just say Number of Mel-frequency bins here... triangular corresponds to the filter being used. OR `Number of Mel-frequency bins from triangular filterbank`

patrickvonplaten commented on src/transformers/models/speech_to_text/feature_extraction_speech_to_text.py:
```diff
@@ -0,0 +1,233 @@
+        num_mel_bins (:obj:`int`, defaults to 80):
+            Number of triangular Mel-frequency bins
+        padding_value (:obj:`float`, defaults to 0.0):
+            The value that is used to fill the padding values.
```
```suggestion
            The value that is used to fill the padding vectors.
```

patrickvonplaten commented on src/transformers/models/speech_to_text/feature_extraction_speech_to_text.py:
```diff
@@ -0,0 +1,233 @@
+
+                This is especially useful to enable the use of Tensor Cores on NVIDIA hardware with compute capability
+                >= 7.5 (Volta), or on TPUs which benefit from having sequence lengths be a multiple of 128.
+            return_attention_mask (:obj:`bool`, `optional`):
```
I'd leave it here, but not make it part of the config...

patrickvonplaten replied:
So remove it from init

patrickvonplaten commented on src/transformers/models/speech_to_text/feature_extraction_speech_to_text.py:
```diff
@@ -0,0 +1,233 @@
+            sampling_rate (:obj:`int`, `optional`):
+                The sampling rate at which the ``raw_speech`` input was sampled. It is strongly recommended to pass
+                ``sampling_rate`` at the forward call to prevent silent errors.
+            padding_value (:obj:`float`, defaults to 0.0):
```
this a short line below this one is missing

sgugger reviewed (commented):
I have one main issue with the objects added to the main init where we need to take more care. Specifically:

- `Speech2TextTokenizer` requires sentencepiece, so it should only be imported in an `if is_sentencepiece_available()` block.
- `Speech2TestFeatureExtractor` requires torch and torchaudio, but both of those are imported in its file inside proper tests, so the object can actually be available at all time. It just needs a check at init that will return an ImportError if torch and torchaudio are not available.
- `Speech2TextProcessor` then only requires sentencepiece and should be imported in an `if is_sentencepiece_available` block

Once this is sorted out, my main comments regards the docstrings. Remember rst is not Markdown :-)

patrickvonplaten commented on src/transformers/models/speech_to_text/modeling_speech_to_text.py:
```diff
@@ -0,0 +1,1348 @@
+    via gated linear units (https://arxiv.org/abs/1911.08460)
+    """
+
+    def __init__(
```
reformat by hand here - think this should be in one line

patrickvonplaten commented on src/transformers/models/speech_to_text/modeling_speech_to_text.py:
```diff
@@ -0,0 +1,1348 @@
+        self.fc2 = nn.Linear(config.encoder_ffn_dim, self.embed_dim)
+        self.final_layer_norm = nn.LayerNorm(self.embed_dim)
+
+    def forward(
```
Can we maybe use a `# Copied from` here?

patrickvonplaten commented on src/transformers/models/speech_to_text/modeling_speech_to_text.py:
```diff
@@ -0,0 +1,1348 @@
+
+SPEECH_TO_TEXT_INPUTS_DOCSTRING = r"""
+    Args:
+        input_features (:obj:`torch.LongTensor` of shape :obj:`(batch_size, fbank_feature_length, num_mel_bins)`):
```
think I'd prefer to use `seq_length` and `feature_size` here instead since we use these names in the feature extractor as well

patrickvonplaten commented on src/transformers/models/speech_to_text/modeling_speech_to_text.py:
```diff
@@ -0,0 +1,1348 @@
+    ):
+        r"""
+        Args:
+            input_features (:obj:`torch.LongTensor` of shape :obj:`(batch_size, fbank_feature_length, num_mel_bins)`):
```
same here

patrickvonplaten commented on src/transformers/models/speech_to_text/processing_speech_to_text.py:
```diff
@@ -0,0 +1,144 @@
+
+    :class:`~transformers.Speech2TextProcessor` offers all the functionalities of
+    :class:`~transformers.Speech2TextFeatureExtractor` and :class:`~transformers.Speech2TextTokenizer`. See the
+    docstring of :meth:`~transformers.Speech2TextProcessor.__call__` and
```
```suggestion
    :meth:`~transformers.Speech2TextProcessor.__call__` and
```

patrickvonplaten commented on src/transformers/models/speech_to_text/processing_speech_to_text.py:
```diff
@@ -0,0 +1,144 @@
+
+    def save_pretrained(self, save_directory):
+        """
+        Save a Speech2Text feature_extractor object and Speech2Text tokenizer object to the directory
```
```suggestion
        Save a Speech2Text feature extractor object and Speech2Text tokenizer object to the directory
```

patrickvonplaten reviewed (approved):
Very nice overall! I mainly left name change suggestions for the feature extractor and I'd prefer to not put the model into `AutoModelForSeq2SeqLM`. Apart from this and @sgugger comments, it looks good to me.

LysandreJik commented on src/transformers/models/speech_to_text/configuration_speech_to_text.py:
```diff
@@ -0,0 +1,200 @@
+# coding=utf-8
+# Copyright 2021 The HuggingFace Inc. team. All rights reserved.
```
Should we mention Facebook in the copyright?

LysandreJik commented on src/transformers/models/speech_to_text/configuration_speech_to_text.py:
```diff
@@ -0,0 +1,200 @@
+        self.input_channels = input_channels
+
+        if len(self.conv_kernel_sizes) != self.num_conv_layers:
+            raise ValueError(
+                "Configuration for convolutional module is incorrect."
+                "It is required that `len(config.conv_kernel_sizes)` == `config.num_conv_layers`"
+                f"but is `len(config.conv_kernel_sizes) = {len(self.conv_kernel_sizes)}`,"
+                f"`config.num_conv_layers = {self.num_conv_layers}`."
+            )
```
Very nice error message

LysandreJik reviewed (approved):
This is very clean, fantastic work @patil-suraj!

I may be wrong, but I haven't seen the instructions to download libsndfile1-dev, which I think is important.
The instruction you've put in the CI is very debian-specific, so I think it would be nice to document the exact thing we're trying to install, rather than a specific instruction, for example:

S2T requires the development version of the [libsndfile](http://www.mega-nerd.com/libsndfile/) package which can be installed via a system package manager. On Ubuntu it can be installed as follows:
```py
apt install libsndfile1-dev
```

Other than that, LGTM!

patil-suraj commented:
I've added proper instructions to install the extra dependencies and addressed Patrick and Sylvain's comments regarding the docs and imports. All slow/non-slow tests are passing!

Merging!

xjdeng commented:
Hey,

Sorry, I'm still a beginner compared to the rest of the folks here so sorry if my question is a little basic.

But I'm trying to build a pipeline to manually transcribe Youtube videos (that aren't transcribed correctly by Google) and I was considering using your [model ](https://huggingface.co/facebook/s2t-small-librispeech-asr)for it.

Here's my unfinished code on  [Google Colab](https://colab.research.google.com/drive/15SgSw1KmD-sxdf6Zd953fJIPSmisHd6M?usp=sharing); the last line throws an error:

```
!pip install git+https://github.com/huggingface/transformers
!pip install youtube-dl path.py soundfile librosa sentencepiece torchaudio

import youtube_dl
from path import Path as Path
import tempfile
import textwrap
import librosa
import soundfile as sf
import torch
from transformers import Speech2TextProcessor, Speech2TextForConditionalGeneration

model = Speech2TextForConditionalGeneration.from_pretrained("facebook/s2t-small-librispeech-asr")
processor = Speech2TextProcessor.from_pretrained("facebook/s2t-small-librispeech-asr")
wrapper = textwrap.TextWrapper(width=70)

mydir = tempfile.TemporaryDirectory()
dirname = mydir.name + "/tmp.wav"

!youtube-dl -o $dirname -ci -f 'bestvideo[ext=mp4]+bestaudio' -x --audio-format wav https://www.youtube.com/watch?v=d5yfUuHYWho

filename = dirname + ".wav"

speech, rate = sf.read(filename)
speech = librosa.resample(speech.T, rate, 16000)

features = processor(speech, sampling_rate=16000, padding=True, return_tensors="pt")
```

And here's the error produced:

```
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-8-8fc3e2d943e0> in <module>()
----> 1 features = processor(speech, sampling_rate=16000, padding=True, return_tensors="pt")

5 frames
/usr/local/lib/python3.7/dist-packages/torchaudio/compliance/kaldi.py in _get_waveform_and_window_properties(waveform, channel, sample_frequency, frame_shift, frame_length, round_to_power_of_two, preemphasis_coefficient)
    147     assert 2 <= window_size <= len(
    148         waveform), ('choose a window size {} that is [2, {}]'
--> 149                     .format(window_size, len(waveform)))
    150     assert 0 < window_shift, '`window_shift` must be greater than 0'
    151     assert padded_window_size % 2 == 0, 'the padded `window_size` must be divisible by two.' \

AssertionError: choose a window size 400 that is [2, 2]
```

Can anyone point me in the right direction?  Thanks.

patil-suraj commented:
hi @xjdeng

Thanks for reporting this. This PR is now merged. So could you please open an issue with this error, we will discuss it there.
